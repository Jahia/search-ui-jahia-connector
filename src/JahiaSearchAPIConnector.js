import request from './request';
import adaptRequest from './requestAdapter';
import adaptResponse from './responseAdapter';
import Constants from './constants';

class JahiaSearchAPIConnector {
    /**
     * @typedef Options
     * @param  {string} apiToken Credential found in your Jahia Tools
     * @param  {string} baseURL  URL of your Jahia installation
     * @param  {string} siteKey The site search will be performed in
     * @param  {string} language Language in which search will be performed
     * @param  {string} workspace Workspace in which search will be performed
     * @param  {string} nodeType The node type that should be searched for
     */
    constructor({
        apiToken,
        baseURL,
        siteKey,
        language = Constants.LANGUAGE,
        workspace = Constants.WORKSPACE,
        nodeType = Constants.NODE_FILTER.TYPE
    }) {
        this.apiToken = apiToken;
        this.baseURL = baseURL;
        this.siteKey = siteKey;
        this.language = language;
        this.workspace = workspace;
        this.nodeType = nodeType;
    }

    onSearch(state, queryConfig) {
        console.log("state",state,"query", queryConfig);
        let requestOptions = {
            siteKey: this.siteKey,
            language: this.language,
            workspace: this.workspace,
            nodeType: this.nodeType
        };
        const query = adaptRequest(requestOptions, state, queryConfig);
        return request(this.apiToken, this.baseURL, 'POST', query).then(json =>
            adaptResponse(json, state.resultsPerPage)
        );
    }

    async onAutocomplete({searchTerm}, queryConfig) {
        if (queryConfig.suggestions) {
            console.warn(
                'search-ui-jahia-connector: Site Search does support query suggestions on autocomplete'
            );
        }
        if (queryConfig.results) {
            let requestOptions = {
                siteKey: this.siteKey,
                language: this.language,
                workspace: this.workspace,
                nodeType: this.nodeType
            };
            const query = adaptRequest(requestOptions,
                {searchTerm},
                queryConfig.results
            );

            return request(this.apiToken, this.baseURL, 'POST', query).then(json => ({
                autocompletedResults: adaptResponse(json, queryConfig.results.resultsPerPage).results
            }));
        }
    }

    onAutocompleteResultClick({ query, documentId, tags }) {
        if (tags) {
            console.warn(
                "search-ui-jahia-connector: Site Search does not support tags on autocompleteClick"
            );
        }
        //Todo integrate with JCustomer
    }
}

export default JahiaSearchAPIConnector;
