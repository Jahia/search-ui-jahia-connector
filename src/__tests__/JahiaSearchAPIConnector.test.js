import JahiaSearchAPIConnector from '..';

import exampleAPIResponse from '../../resources/example-response.json';

function fetchResponse(response) {
    return Promise.resolve({
        status: 200,
        json: () => Promise.resolve(response)
    });
}

beforeEach(() => {
    global.Headers = jest.fn();
    global.fetch = jest.fn().mockReturnValue(fetchResponse(exampleAPIResponse));
});

const apiToken = 12345;
const baseURL = 'http://localhost:8080';
const siteKey = 'localhost';

const params = {
    apiToken,
    baseURL,
    siteKey
};

it('can be initialized', () => {
    const connector = new JahiaSearchAPIConnector(params);
    expect(connector).toBeInstanceOf(JahiaSearchAPIConnector);
});

describe('#onSearch', () => {
    function subject({state, queryConfig = {}}) {
        const connector = new JahiaSearchAPIConnector({
            ...params
        });
        return connector.onSearch(state, queryConfig);
    }

    it('will correctly format an API response', async () => {
        const response = await subject({state: {}, queryConfig: {}});
        expect(response).toMatchSnapshot();
    });
});

describe('#onAutocomplete', () => {
    function subject({
        state,
        queryConfig = {
            results: {}
        }
    }) {
        const connector = new JahiaSearchAPIConnector({
            ...params
        });
        return connector.onAutocomplete(state, queryConfig);
    }

    it('will correctly format an API response', async () => {
        const response = await subject({
            state: {},
            queryConfig: {
                results: {}
            }
        });
        expect(response).toMatchSnapshot();
    });

    it('will not return anything for suggestions', async () => {
        const response = await subject({
            state: {},
            queryConfig: {
                suggestions: {}
            }
        });
        expect(response).toMatchSnapshot();
    });
});

describe("#onAutocompleteResultClick", () => {
    function subject(clickData) {
        const connector = new JahiaSearchAPIConnector(params);
        return connector.onAutocompleteResultClick(clickData);
    }

    it("will call the API with the correct body params", async () => {
        const query = "test";
        const documentId = "12345";
        const tags = "12345";

        const response = await subject({
            query,
            documentId,
            tags
        });

        expect(response).toMatchSnapshot();
    });
});
