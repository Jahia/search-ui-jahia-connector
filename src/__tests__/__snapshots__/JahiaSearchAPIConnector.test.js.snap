// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#onAutocomplete will correctly format an API response 1`] = `
Object {
  "autocompletedResults": Array [
    Object {
      "excerpt": Object {
        "snippet": "Jahia <em>cluster</em> <em>setup</em> ... The Jahia Team Jahia <em>cluster</em> <em>setup</em> Here are the steps followed to install an Jahia <em>cluster</em> for a production environment. We will also detail",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/developer/dx/techwiki/misc/dx-cluster-setup.html",
      },
      "title": Object {
        "raw": "Jahia cluster setup",
      },
      "updatedAt": Object {
        "raw": "Tue Jul 16 04:07:26 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "ElasticSearch / Unomi <em>cluster</em> <em>setup</em> ... The Jahia Team ElasticSearch / Unomi <em>cluster</em> <em>setup</em> Please note that the following instructions are not 100% valid for JExperience 1.10.1 and Apache",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/developer/dx/techwiki/misc/elasticsearch--unomi-cluster-setup.html",
      },
      "title": Object {
        "raw": "ElasticSearch / Unomi cluster setup",
      },
      "updatedAt": Object {
        "raw": "Wed Jul 17 03:39:23 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "6 - <em>Cluster</em> <em>setup</em> ... The Jahia Team 6 - <em>Cluster</em> <em>setup</em> 6.1 Discovery

6.1.1 Overview

Discovery must be configured on every layer of the <em>cluster</em> architecture",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/legacy-1/marketing-factory/18/technical/installation-configuration-and-f/6---cluster-setup.html",
      },
      "title": Object {
        "raw": "6 - Cluster setup",
      },
      "updatedAt": Object {
        "raw": "Tue Jul 24 08:46:50 EDT 2018",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": " during the Digital Experience Manager configuration wizard (<em>cluster</em> installation), please restore it at the appropriate location.

Remove the 2 “lock ...  (DX 7.2.1.1+)

The safe backup restore here is only relevant, when you are restoring a DX <em>clustered</em> backup at another infrastructure, say cloning",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/system-administrator/dx/72/technical/configuration-and-fine-tuning/faq.html",
      },
      "title": Object {
        "raw": "FAQs",
      },
      "updatedAt": Object {
        "raw": "Tue Apr 16 14:03:29 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "
|_________indexing_configuration_version.xml

If you have chosen an alternative location of the datastore folder during the Jahia configuration wizard (<em>cluster</em> installation ...  keep the former indexes to save time.

Safe backup restore

The safe backup restore here is only relevant, when you are restoring a Jahia <em>clustered</em>",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/system-administrator/dx/73/technical/configuration-and-fine-tuning/faq.html",
      },
      "title": Object {
        "raw": "FAQs",
      },
      "updatedAt": Object {
        "raw": "Mon Jul 15 10:38:44 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": " during the Digital Experience Manager configuration wizard (<em>cluster</em> installation), please restore it at the appropriate location.

Remove the 2 “lock ...  of a restored DX instance.

It includes currently:


 the purge of the JGROUPSPING database table data, which is responsible for the <em>cluster</em> membership",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/trash/dx-new-docs-do-not-suppress/technical/configuration-and-fine-tuning/faq.html",
      },
      "title": Object {
        "raw": "FAQ",
      },
      "updatedAt": Object {
        "raw": "Mon Feb 19 04:18:00 EST 2018",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": " for <em>clustering</em> settings. Since Digital Experience Manager 7.0.0.2 the runtime data and configuration files are located outside the Web application ...  features and services.
 jahia.node.properties – configuration file for the <em>clustering</em> options
 license.xml – is a file with your current Digital",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/trash/dx-new-docs-do-not-suppress/technical/configuration-and-fine-tuning/configuring.html",
      },
      "title": Object {
        "raw": "Configuring",
      },
      "updatedAt": Object {
        "raw": "Mon Feb 19 04:18:05 EST 2018",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": " configuration options are gathered in the jahia.properties and jahia.node.properties files. The latter one is responsible only for <em>clustering</em> settings. Since ...  and services.
 jahia.node.properties – configuration file for the <em>clustering</em> options
 license.xml – is a file with your current Digital Experience Manager",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/system-administrator/dx/72/technical/configuration-and-fine-tuning/configuring.html",
      },
      "title": Object {
        "raw": "Configuring DX features",
      },
      "updatedAt": Object {
        "raw": "Tue Apr 16 14:02:29 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": " are gathered in the jahia.properties and jahia.node.properties files. The latter one is responsible only for <em>clustering</em> settings. The runtime data ...  and services.
 jahia.node.properties – configuration file for the <em>clustering</em> options
 license.xml – is a file with your current Jahia license",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/system-administrator/dx/73/technical/configuration-and-fine-tuning/configuring.html",
      },
      "title": Object {
        "raw": "Configuring Jahia features",
      },
      "updatedAt": Object {
        "raw": "Tue Jul 16 04:07:26 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/trash/dx-new-docs-do-not-suppress/development/development-best-practices.html",
      },
      "title": Object {
        "raw": "Development Best Practices",
      },
      "updatedAt": Object {
        "raw": "Mon Feb 19 04:18:05 EST 2018",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/developer/dx/72/development/development-best-practices.html",
      },
      "title": Object {
        "raw": "Development best practices",
      },
      "updatedAt": Object {
        "raw": "Tue Apr 16 14:12:46 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/developer/dx/73/development/development-best-practices.html",
      },
      "title": Object {
        "raw": "Development best practices",
      },
      "updatedAt": Object {
        "raw": "Tue Jul 16 08:36:55 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "AELB-26 qa At installation time, Jahia generate a UUID to ensure that each <em>cluster</em> node id is unique.
From https://wiki.apache.org/jackrabbit ... /<em>Clustering</em>#Unique_Cluster_Node_ID it is safe to use any kind of identifier for this property as long as this ID is unique on each <em>cluster</em> nodes. It could",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/training--kb/how-to.html",
      },
      "title": Object {
        "raw": "Knowledge Base",
      },
      "updatedAt": Object {
        "raw": "Tue Aug 06 09:59:46 EDT 2019",
      },
    },
  ],
}
`;

exports[`#onAutocomplete will not return anything for suggestions 1`] = `undefined`;

exports[`#onAutocompleteResultClick will call the API with the correct body params 1`] = `undefined`;

exports[`#onSearch will correctly format an API response 1`] = `
Object {
  "requestId": "",
  "results": Array [
    Object {
      "excerpt": Object {
        "snippet": "Jahia <em>cluster</em> <em>setup</em> ... The Jahia Team Jahia <em>cluster</em> <em>setup</em> Here are the steps followed to install an Jahia <em>cluster</em> for a production environment. We will also detail",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/developer/dx/techwiki/misc/dx-cluster-setup.html",
      },
      "title": Object {
        "raw": "Jahia cluster setup",
      },
      "updatedAt": Object {
        "raw": "Tue Jul 16 04:07:26 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "ElasticSearch / Unomi <em>cluster</em> <em>setup</em> ... The Jahia Team ElasticSearch / Unomi <em>cluster</em> <em>setup</em> Please note that the following instructions are not 100% valid for JExperience 1.10.1 and Apache",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/developer/dx/techwiki/misc/elasticsearch--unomi-cluster-setup.html",
      },
      "title": Object {
        "raw": "ElasticSearch / Unomi cluster setup",
      },
      "updatedAt": Object {
        "raw": "Wed Jul 17 03:39:23 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "6 - <em>Cluster</em> <em>setup</em> ... The Jahia Team 6 - <em>Cluster</em> <em>setup</em> 6.1 Discovery

6.1.1 Overview

Discovery must be configured on every layer of the <em>cluster</em> architecture",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/legacy-1/marketing-factory/18/technical/installation-configuration-and-f/6---cluster-setup.html",
      },
      "title": Object {
        "raw": "6 - Cluster setup",
      },
      "updatedAt": Object {
        "raw": "Tue Jul 24 08:46:50 EDT 2018",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": " during the Digital Experience Manager configuration wizard (<em>cluster</em> installation), please restore it at the appropriate location.

Remove the 2 “lock ...  (DX 7.2.1.1+)

The safe backup restore here is only relevant, when you are restoring a DX <em>clustered</em> backup at another infrastructure, say cloning",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/system-administrator/dx/72/technical/configuration-and-fine-tuning/faq.html",
      },
      "title": Object {
        "raw": "FAQs",
      },
      "updatedAt": Object {
        "raw": "Tue Apr 16 14:03:29 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "
|_________indexing_configuration_version.xml

If you have chosen an alternative location of the datastore folder during the Jahia configuration wizard (<em>cluster</em> installation ...  keep the former indexes to save time.

Safe backup restore

The safe backup restore here is only relevant, when you are restoring a Jahia <em>clustered</em>",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/system-administrator/dx/73/technical/configuration-and-fine-tuning/faq.html",
      },
      "title": Object {
        "raw": "FAQs",
      },
      "updatedAt": Object {
        "raw": "Mon Jul 15 10:38:44 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": " during the Digital Experience Manager configuration wizard (<em>cluster</em> installation), please restore it at the appropriate location.

Remove the 2 “lock ...  of a restored DX instance.

It includes currently:


 the purge of the JGROUPSPING database table data, which is responsible for the <em>cluster</em> membership",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/trash/dx-new-docs-do-not-suppress/technical/configuration-and-fine-tuning/faq.html",
      },
      "title": Object {
        "raw": "FAQ",
      },
      "updatedAt": Object {
        "raw": "Mon Feb 19 04:18:00 EST 2018",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": " for <em>clustering</em> settings. Since Digital Experience Manager 7.0.0.2 the runtime data and configuration files are located outside the Web application ...  features and services.
 jahia.node.properties – configuration file for the <em>clustering</em> options
 license.xml – is a file with your current Digital",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/trash/dx-new-docs-do-not-suppress/technical/configuration-and-fine-tuning/configuring.html",
      },
      "title": Object {
        "raw": "Configuring",
      },
      "updatedAt": Object {
        "raw": "Mon Feb 19 04:18:05 EST 2018",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": " configuration options are gathered in the jahia.properties and jahia.node.properties files. The latter one is responsible only for <em>clustering</em> settings. Since ...  and services.
 jahia.node.properties – configuration file for the <em>clustering</em> options
 license.xml – is a file with your current Digital Experience Manager",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/system-administrator/dx/72/technical/configuration-and-fine-tuning/configuring.html",
      },
      "title": Object {
        "raw": "Configuring DX features",
      },
      "updatedAt": Object {
        "raw": "Tue Apr 16 14:02:29 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": " are gathered in the jahia.properties and jahia.node.properties files. The latter one is responsible only for <em>clustering</em> settings. The runtime data ...  and services.
 jahia.node.properties – configuration file for the <em>clustering</em> options
 license.xml – is a file with your current Jahia license",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/system-administrator/dx/73/technical/configuration-and-fine-tuning/configuring.html",
      },
      "title": Object {
        "raw": "Configuring Jahia features",
      },
      "updatedAt": Object {
        "raw": "Tue Jul 16 04:07:26 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/trash/dx-new-docs-do-not-suppress/development/development-best-practices.html",
      },
      "title": Object {
        "raw": "Development Best Practices",
      },
      "updatedAt": Object {
        "raw": "Mon Feb 19 04:18:05 EST 2018",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/developer/dx/72/development/development-best-practices.html",
      },
      "title": Object {
        "raw": "Development best practices",
      },
      "updatedAt": Object {
        "raw": "Tue Apr 16 14:12:46 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/documentation/developer/dx/73/development/development-best-practices.html",
      },
      "title": Object {
        "raw": "Development best practices",
      },
      "updatedAt": Object {
        "raw": "Tue Jul 16 08:36:55 EDT 2019",
      },
    },
    Object {
      "excerpt": Object {
        "snippet": "AELB-26 qa At installation time, Jahia generate a UUID to ensure that each <em>cluster</em> node id is unique.
From https://wiki.apache.org/jackrabbit ... /<em>Clustering</em>#Unique_Cluster_Node_ID it is safe to use any kind of identifier for this property as long as this ID is unique on each <em>cluster</em> nodes. It could",
      },
      "id": Object {
        "raw": undefined,
      },
      "link": Object {
        "raw": "http://localhost/cms/render/default/en/sites/academy/home/training--kb/how-to.html",
      },
      "title": Object {
        "raw": "Knowledge Base",
      },
      "updatedAt": Object {
        "raw": "Tue Aug 06 09:59:46 EDT 2019",
      },
    },
  ],
  "totalPages": NaN,
  "totalResults": 13,
}
`;
